contracts/adapters/swap/shade-protocol/src/contract.rs

```diff
@@ -0,0 +1,435 @@
+        .add_attribute("action", "migrate")
+        .add_attribute(
+            "entry_point_contract_address",
+            checked_entry_point_contract.address.to_string(),
+        )
+        .add_attribute(
+            "shade_router_contract_address",
+            checked_shade_router_contract.address.to_string(),
+        ))
+}
+
+///////////////////
+/// INSTANTIATE ///
+///////////////////
+
+#[cfg_attr(not(feature = "library"), entry_point)]
+pub fn instantiate(
+    deps: DepsMut,
+    _env: Env,
+    _info: MessageInfo,
+    msg: InstantiateMsg,
+) -> ContractResult<Response> {
+    // Set contract version
+    // set_contract_version(deps.storage, CONTRACT_NAME, CONTRACT_VERSION)?;
+
+    // Validate entry point contract
+    let checked_entry_point_contract = ContractInfo {
+        address: deps
+            .api
+            .addr_validate(&msg.entry_point_contract.address.to_string())?,
+        code_hash: msg.entry_point_contract.code_hash,
+    };
+
+    ENTRY_POINT_CONTRACT.save(deps.storage, &checked_entry_point_contract)?;
+
+    let checked_shade_router_contract = ContractInfo {
+        address: deps
+            .api
+            .addr_validate(&msg.shade_router_contract.address.to_string())?,
+        code_hash: msg.shade_router_contract.code_hash,
+    };
+
+    SHADE_ROUTER_CONTRACT.save(deps.storage, &checked_shade_router_contract)?;
+
+    VIEWING_KEY.save(deps.storage, &msg.viewing_key)?;
+
+    Ok(Response::new()
+        .add_attribute("action", "instantiate")
+        .add_attribute(
+            "entry_point_contract_address",
+            checked_entry_point_contract.address.to_string(),
+        )
+        .add_attribute(
+            "shade_router_contract_address",
+            msg.shade_router_contract.address,
+        ))
+}
+
+///////////////
+/// RECEIVE ///
+///////////////
+
+// Receive is the main entry point for the contract to
+// receive cw20 tokens and execute the swap
+pub fn receive_snip20(
+    deps: DepsMut,
+    env: Env,
+    mut info: MessageInfo,
+    snip20_msg: Snip20ReceiveMsg,
+) -> ContractResult<Response> {
+    // Set the sender to the originating address that triggered the cw20 send call
+    // This is later validated / enforced to be the entry point contract address
+    info.sender = deps.api.addr_validate(&snip20_msg.sender.to_string())?;
+
+    match snip20_msg.msg {
+        Some(msg) => match from_binary(&msg)? {
+            Snip20HookMsg::Swap { operations } => {
+                execute_swap(deps, env, info, operations, snip20_msg.amount)
+            }
+        },
+        None => Ok(Response::default()),
+    }
+}
+
+///////////////
+/// EXECUTE ///
+///////////////
+
+#[cfg_attr(not(feature = "library"), entry_point)]
+pub fn execute(
+    deps: DepsMut,
+    env: Env,
+    info: MessageInfo,
+    msg: ExecuteMsg,
+) -> ContractResult<Response> {
+    match msg {
+        ExecuteMsg::Receive(snip20_msg) => receive_snip20(deps, env, info, snip20_msg),
```

Comment:
This is because Shade Protocol only supports snip20 tokens, right?

Reply:
Correct, the only tokens coming in will be from the receive TX called by the sent snip20